

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>History and Future Directions &mdash; Medaka 0.0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Medaka 0.0.1 documentation" href="index.html"/>
        <link rel="next" title="medaka package" href="medaka.html"/>
        <link rel="prev" title="Medaka Examples" href="examples.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Medaka
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="examples.html">Medaka Examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">History and Future Directions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="medaka.html">medaka package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Medaka</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>History and Future Directions</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/future.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="history-and-future-directions">
<span id="futuredirections"></span><h1>History and Future Directions<a class="headerlink" href="#history-and-future-directions" title="Permalink to this headline">¶</a></h1>
<p>Medaka has grown out of some early work on the study of errors in basecalls
from recurrent neural network (RNN) basecallers of Oxford Nanopore
Technologies’ (ONT) data. It is relatively straightforward to enumerate
dominant error classes in reads and correct for some of these. For example it
is well known that deletion of bases forms a large proportion of all errors;
occurring both within and outside homopolymers. Perhaps not so well known is
that these deletions can show a ‘strand’ bias, being present in one orientation
of reads but not the other. This fact can be exploited. Results from ad-hoc,
aggressive repairing of deletions was presented by ONT at London Calling 2017.</p>
<p>Having corrected dominant error classes it becomes progressively harder to
achieve further improvement by manual inspection. At this point persuing
automated methods to learn and correct for errors becomes prudent. As a first
attempt Medaka implements an interface to prepare training data from alignment
of basecalls and a truth set to a common baseline. This baseline may be
anything, perhaps a draft assembly or chosen single molecule basecalls. Once
these data have been collated they may be used directly or prepared further to
exploit known errors. For example one may reduce or augment the with data counts
of orientated bases and indels as this is known to be a relavant consideration.</p>
<p>The framework provided has had limited testing and benchmarking, but it is
being released in an effort to allow researchers to experiment without having
to write much of the tedious data preparation code. Researchers can extend the
tools provided to achieve better results that those obtained currently. There
are certainly obvious extension which can be made, including:</p>
<ul class="simple">
<li>use of quality scores from the basecaller,</li>
<li>relax contraints on how features are selected: there are currently some
shortcuts taken which could better handle alignment subtleties,</li>
<li>modifying the feature vectors to take wider windows of data,</li>
<li>altering the size, structure, and connections between the RNN layers</li>
<li>use of convolutional layers.</li>
</ul>
<p>Using lower level data directly (not basecalls alone) will provide a more
powerful method; this is something which is being actively researched. Early
examples include <a class="reference external" href="https://github.com/jts/nanopolish">nanopolish</a> and
<a class="reference external" href="https://github.com/tszalay/poreseq">poreseq</a> which take the approach of
iteratively refining a candidate sequence by alignment of nanopore event data
to the candidate sequence. Where their data models predict discrepancies
between the observed data and the candidate sequence, these methods mutate the
candidate to achieve better alignment scores. In this way the methods maximise
the probability of having observed all input event data under their model.</p>
<p>Neural network classifiers have been shown (in the context of basecalling) to
model better the primary raw, and secondary event, data from nanopore devices
than the generative Hidden Markov Models used in both of the above methods.
It is natural therefore to desire to encorporate neural networks into a
procedure to error correct candidate sequences from primary and secondary
data; this is the focus of current research. Further it may be possible that
such an approach need not be iterative: a single pass on the inputs could be
performed to achieve results in quicker time.</p>
<p>A future release on medaka will encorporate these higher-order methods.</p>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="medaka.html" class="btn btn-neutral float-right" title="medaka package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="examples.html" class="btn btn-neutral" title="Medaka Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Oxford Nanopore Technologies.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>